src/api.ts

```diff
@@ -0,0 +1,74 @@
+import {
+  AbstractDetectorDict,
+  AbstractSourceDict,
+  BotdError,
+  BotDetectionResult,
+  BotKind,
+  Component,
+  ComponentDict,
+  DetectionDict,
+  State,
+} from './types'
+
+export function detect<T extends AbstractDetectorDict>(
+  components: ComponentDict<any>,
+  detectors: T,
+): [DetectionDict<T>, BotDetectionResult] {
```

Comment:
I'd try to describe the dependency between the components' and detectors' types in TypeScript (the detectors expect a specific type of component dictionary)